third_party/move/tools/move-linter/src/model_ast_lints/nonminimal_bool.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module implements an expression linter that checks for boolean identities
+//! that occur in conditionals:
+//! 1. `x && true`, which can be replaced with just `x`.
+//! 2. `x && false`, which can be replaced with just `false`.
+//! 3. `x || true`, which can be replaced with just `true`.
+//! 4. `x || false`, which can be replaced with just `x`.
+//! 5. `x <==> true`, which can be replaced with just `x`.
+//! 6. `x <==> false`, which can be replaced with just `!x`.
+//! 7. `x ==> true`, which can be replaced with just `true`.
+//! 8. `x ==> false`, which can be replaced with just `!x`.
+//! 9. `true ==> x`, which can be replaced with just `x`.
+//! 10. `false ==> x`, which can be replaced with just `true`.
+//! 11. `!true`, which can be replaced with just `false`.
+//! 12. `!false`, which can be replaced with just `true`.
+//!
+//! Note also that rules 1 through 6 have both LHS and RHS version
+
+use move_compiler_v2::external_checks::ExpChecker;
+use move_model::{
+    ast::{ExpData, Operation, Value},
+    model::GlobalEnv,
+};
+
+#[derive(Default)]
+pub struct NonMinimalBool;
+
+impl ExpChecker for NonMinimalBool {
+    fn get_name(&self) -> String {
+        "nonminimal_bool".to_string()
+    }
+
+    fn visit_expr_pre(&mut self, env: &GlobalEnv, expr: &ExpData) {
+        use ExpData::{Call, Value as ExpValue};
+        use Operation::*;
+        use Value::Bool;
+
+        let get_msg = |cmp: &Operation, b: bool, side| {
+            match (cmp, b, side) {
+            (And, true, s) | (Or, false, s) => Some(format!(
+                "The {} occurrence of `{}` is redundant and should be removed.",
```

Comment:
```suggestion
                "The {} occurrence of `{}` is redundant and could be removed.",
```

Reply:
Also, for the code:

```
    const DEBUG: bool = true;

    fun test_mut() {
        DEBUG && false;
    }
```

We say:

```
+ warning: [lint] The left occurrence of `true` is redundant and should be removed.
+    ┌─ tests/model_ast_lints/nonminimal_bool.move:53:9
+    │
+ 53 │         DEBUG && false;
+    │         ^^^^^^^^^^^^^^
+    │
```

Constants are already evaluated by this point, and so `true` may not occur literally. The warning text could be adjusted to say something like: "the left-hand-side of `&&` evaluates to `true`, therefore the whole expression could be simplified".

Reply:
How generic should I go with suggestion message? The current suggestions give the user the exact reduction to apply (true && x is x, etc), but if we just want to say "the {direction}-hand-side of {operator} evaluates to {bool literal}, therefore the whole expression could be simplified", then this general pattern can just be used everywhere.

Reply:
We can provide specific messages, if it is correct.

We should not, IMO, provide wrong but specific messages (I would not say the message in this specific context is wrong, but was trying to showcase a real-world scenario with constants - we have code like this in our Rust code base).

If getting correct specific messages has diminishing returns then might not be worth the complexity.

Reply:
This makes sense. Will keep this in mind for future (especially as i plan to write a few more lint rules soon). Sounds like we're in agreement about current error message being fine.